presto-orc/src/main/java/com/facebook/presto/orc/reader/MapFlattenedStreamReader.java

```diff
@@ -0,0 +1,383 @@
+{
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(MapFlattenedStreamReader.class).instanceSize();
+
+    private final StreamDescriptor streamDescriptor;
+    private final DateTimeZone hiveStorageTimeZone;
+    private final AggregatedMemoryContext systemMemoryContext;
+    private final StreamDescriptor baseValueStreamDescriptor;
+    private final OrcType.OrcTypeKind keyOrcType;
+
+    private final List<InputStreamSource<BooleanInputStream>> inMapStreamSources = new ArrayList<>();
+    private final List<BooleanInputStream> inMapStreams = new ArrayList<>();
+    private final List<StreamReader> valueStreamReaders = new ArrayList<>();
+    private final List<StreamDescriptor> valueStreamDescriptors = new ArrayList<>();
+    private final List<DwrfProto.KeyInfo> mapKeys = new ArrayList<>();
+
+    private int readOffset;
+    private int nextBatchSize;
+
+    @Nonnull
+    private InputStreamSource<BooleanInputStream> presentStreamSource = missingStreamSource(BooleanInputStream.class);
+    @Nullable
+    private BooleanInputStream presentStream;
+
+    private boolean rowGroupOpen;
+
+    public MapFlattenedStreamReader(StreamDescriptor streamDescriptor, DateTimeZone hiveStorageTimeZone, AggregatedMemoryContext systemMemoryContext)
+    {
+        this.streamDescriptor = requireNonNull(streamDescriptor, "stream is null");
+        this.hiveStorageTimeZone = requireNonNull(hiveStorageTimeZone, "hiveStorageTimeZone is null");
+        this.systemMemoryContext = requireNonNull(systemMemoryContext, "systemMemoryContext is null");
+        this.keyOrcType = streamDescriptor.getNestedStreams().get(0).getStreamType();
+        this.baseValueStreamDescriptor = streamDescriptor.getNestedStreams().get(1);
+    }
+
+    @Override
+    public void prepareNextRead(int batchSize)
+    {
+        readOffset += nextBatchSize;
+        nextBatchSize = batchSize;
+    }
+
+    @Override
+    public Block readBlock(Type type)
+            throws IOException
+    {
+        if (!rowGroupOpen) {
+            openRowGroup();
+        }
+
+        if (readOffset > 0) {
+            if (presentStream != null) {
+                // skip ahead the present bit reader, but count the set bits
+                // and use this as the skip size for the data reader
+                readOffset = presentStream.countBitsSet(readOffset);
+            }
+            if (readOffset > 0) {
+                for (int i = 0; i < valueStreamReaders.size(); i++) {
+                    int valueReadOffset = inMapStreams.get(i).countBitsSet(readOffset);
+                    valueStreamReaders.get(i).prepareNextRead(valueReadOffset);
+                }
+            }
+        }
+
+        // We will use the offsetVector as the buffer to read the length values from lengthStream,
+        // and the length values will be converted in-place to an offset vector.
+        boolean[][] inMapVectors = new boolean[inMapStreams.size()][nextBatchSize];
```

Comment:
If I'm thinking of the PR you're thinking of that was to not initialize nullVector if present stream was null.  Today, DWRF doesn't support suppressing the InMap streams so I think we have to store the data no matter what.  I can change the code to not allocate the inner arrays so eagerly, so we're prepared for one suppression is supported.

Reply:
Nevermind, I found the PR I think you're referencing https://github.com/prestodb/presto/commit/18ebcd93ef83ec9c064bec9b10e9487fe12b27c9

I think inMapVectors is more closely related to offsetVector in arrays.  We need to do one pass to count how many values to read from the underlying readers, then another pass to put them right keys/values in the maps.

Let me think, I'm not sure if it's possible to avoid doing 2 passes.

Reply:
I think the significant part of that PR is to merge the two vector loops to avoid the need for an intermediate array (which can be huge).

Reply:
@kevinwilfong : The idea would be to read and consume the bit read from `inMapStreams.get(i).nextBit()` immediately, so we don't have to store them.

I understand this is not straightforward (and might be infeasible). Feel free to ping me so we can look into together (or try pair-programming :P)